src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -47,145 +54,215 @@ class PortsMatcher(app: AppDefinition, offer: Offer, acceptedResourceRoles: Set[
+    for {
+      resource <- offeredPortsResources
+      rangeInResource <- resource.getRanges.getRangeList.asScala
+      range = Range.inclusive(rangeInResource.getBegin.toInt, rangeInResource.getEnd.toInt)
+    } yield PortRange(resource.getRole, range)
+
+  private[this] def shuffledAvailablePorts: Iterator[PortWithRole] =
+    PortWithRole.randomPortsFromRanges(random)(offeredPortRanges)
+}
+
+object PortsMatcher {
+
+  case class PortWithRole(role: String, port: Int) {
+    def toRange: protos.Range = {
+      protos.Range(port.toLong, port.toLong)
+    }
+  }
+
+  object PortWithRole {
+    /**
+      * Return RangesResources covering all given ports with the given roles.
+      *
+      * Creates as few RangesResources as possible while
+      * preserving the order of the ports.
+      */
+    def createPortsResources(resources: Seq[PortWithRole]): Seq[RangesResource] = {
+      val builder = Seq.newBuilder[RangesResource]
+      @tailrec
+      def process(resources: Seq[PortWithRole]): Unit = resources.headOption match {
+        case None =>
+        case Some(PortWithRole(role, _)) =>
+          val portsForResource: Seq[PortWithRole] = resources.takeWhile(_.role == role)
+          builder += RangesResource(name = Resource.PORTS, createRanges(portsForResource), role = role)
+          process(resources.drop(portsForResource.size))
       }
+      process(resources)
 
-      val byRole: Map[String, Seq[PortMapping]] =
-        mappingsWithAssignedRandoms.groupBy(_.role).mapValues(_.map(_.mapping))
+      builder.result()
+    }
 
-      byRole.map {
-        case (role, portMappings) =>
-          val portRanges = portMappings.map {
-            case PortMapping(containerPort, hostPort, servicePort, protocol) =>
-              protos.Range(hostPort.longValue, hostPort.longValue)
-          }
+    /**
+      * Create as few ranges as possible from the given ports while preserving the order of the ports.
+      *
+      * It does not check if the given ports have different roles.
+      */
+    private[this] def createRanges(ranges: Seq[PortWithRole]): Seq[protos.Range] = {
+      val builder = Seq.newBuilder[protos.Range]
+
+      @tailrec
+      def process(lastRangeOpt: Option[protos.Range], next: Seq[PortWithRole]): Unit = {
+        (lastRangeOpt, next.headOption) match {
+          case (None, _) =>
+          case (Some(lastRange), None) =>
+            builder += lastRange
+          case (Some(lastRange), Some(nextPort)) if lastRange.end == nextPort.port - 1 =>
+            process(Some(lastRange.copy(end = nextPort.port.toLong)), next.tail)
+          case (Some(lastRange), Some(nextPort)) =>
+            builder += lastRange
+            process(Some(nextPort.toRange), next.tail)
+        }
+      }
+      process(ranges.headOption.map(_.toRange), ranges.tail)
+
+      builder.result()
+    }
+
+    /**
+      * We want to avoid reusing the same dynamic ports. Thus we shuffle the ports to make it less likely.
```

Comment:
Short-lived objects are usually no problem for the GC.

I don't understand your suggestion about streams.